arch/arm/core/irq_manage.c

```diff
@@ -115,6 +121,74 @@ void _irq_priority_set(unsigned int irq, unsigned int prio, u32_t flags)
 	NVIC_SetPriority((IRQn_Type)irq, prio);
 }
 
+#elif defined(CONFIG_CPU_CORTEX_R)
+
+/**
+ *
+ * @brief Enable an interrupt line
+ *
+ * Enable the interrupt. After this call, the CPU will receive interrupts for
+ * the specified <irq>.
+ *
+ * @return N/A
+ */
+void _arch_irq_enable(unsigned int irq)
+{
+	struct device *dev = _sw_isr_table[0].arg;
+
+	irq_enable_next_level(dev, (irq >> 8) - 1);
+}
+
+/**
+ *
+ * @brief Disable an interrupt line
+ *
+ * Disable an interrupt line. After this call, the CPU will stop receiving
+ * interrupts for the specified <irq>.
+ *
+ * @return N/A
+ */
+void _arch_irq_disable(unsigned int irq)
+{
+	struct device *dev = _sw_isr_table[0].arg;
+
+	irq_disable_next_level(dev, (irq >> 8) - 1);
+}
+
+/**
+ * @brief Return IRQ enable state
+ *
+ * @param irq IRQ line
+ * @return interrupt enable state, true or false
+ */
+int _arch_irq_is_enabled(unsigned int irq)
+{
+	struct device *dev = _sw_isr_table[0].arg;
+
+	return irq_is_enabled_next_level(dev);
+}
+
+/**
+ * @internal
+ *
+ * @brief Set an interrupt's priority
+ *
+ * The priority is verified if ASSERT_ON is enabled. The maximum number
+ * of priority levels is a little complex, as there are some hardware
+ * priority levels which are reserved: three for various types of exceptions,
+ * and possibly one additional to support zero latency interrupts.
+ *
+ * @return N/A
+ */
+void _irq_priority_set(unsigned int irq, unsigned int prio, u32_t flags)
+{
+	struct device *dev = _sw_isr_table[0].arg;
+
+	irq_set_priority_next_level(dev, (irq >> 8) - 1, prio, flags);
```

Comment:
could you clarify on why it needs "(irq >> 8) -1"?  how about just irq? in our case, we have SPI, which starts from 32, i am not clear on why the code look like it reserve the irq 0 - 256, and let each platform design what irq should be passed in from device tree?

Reply:
In order to reuse isr_wrapper.S when the irq exception is hit, I made the assumption that the GIC (or VIC) will be the first level interrupt controller, so offset 0 of the isr table should have the device and callback of the gic.  The interrupts coming into the GIC are then 2nd level interrupts which use the second nibble of the irq number.  It's not that 0-255 are reserved, it's that the decision on the nesting hierarchy of the interrupt controllers forces that model where the individual interrupt numbers are in the second nibble.

Reply:
Does it mean we assumes that 0xFF is the 1st level? and is it always the case in Zephyr? and thus, for cortex-R5, all the interrupts mask will be 0xFFFFFF00?
Just as R5 doesn't use 1nd level at all, can we  assume  all the interrupt is for 2nd level, and let the conversion to be done inside the GIC controller?

Reply:
The least significant two nibbles will be the first level.  That's always the case for Zephyr.  For the Cortex R cpus, the difference is that the hardware exception table only has one slot for the interrupt handler.  The Cortex R itself does not handle figuring out what peripheral generated the interrupt (unlike Cortex M).  

The Cortex R5 does use 1st level interrupts, it's just that there is only one first level interrupt handler, the GIC.

Note that the GIC is not the only interrupt controller that could be hooked in here.  In my specific case, I am now using an arm pl192 (vic), so I don't think the interrupt controller driver should know anything about the "level" of its irqs.  The number passed into the irq handler should be meaningful only in the frame of reference of the irq handler, not the entire system.  For example, if you wanted to call irq_enable(0x1200) for irq 18 on the gic, my opinion is that the number coming in the parameter to gic_irq_enable should be 18, not 0x1200.

Ideally, the device tree parser should be able to generate the irq tree and handle shifting irq numbers based on where the irq is in the tree, without needing to fixup irq numbers.  But the work is beyond the scope of my knowledge right now.

I did play with trying to put the "shifted" values in the device tree itself, but found that the build system produced a huge isr_sw_table that contained mostly holes of unused irq entries.

Reply:
I have  tried your suggestion to keep the origin value of interrupt in DTS but use the dts fix up to shift the value. It worked for me.